### PR TITLE
Jhbertra/addresses

### PR DESCRIPTION
**Pull request by jhbertra**
Updates the create and redeem endpoints to accept Shelley Addresses. This is a bit of a hack, and support for addresses should be properly pushed upstream into Plutus.

### CODE REVIEW

File: marlowe/src/Language/Marlowe/Client.hs [hunk 1]
```diff
@@ -128,7 +133,7 @@ data PartyAction
              | CloseContract
   deriving (Show)
 
-type RoleOwners = AssocMap.Map Val.TokenName PubKeyHash
+type RoleOwners = AssocMap.Map Val.TokenName (AddressInEra ShelleyEra)
```
Comment (bwbush) [inline marlowe/src/Language/Marlowe/Client.hs:136]: `AddressInEra ShelleyEra` should work. We may want to adjust this eventually, when for instance we allow scripts to hold role tokens.

File: marlowe/src/Language/Marlowe/Client.hs [hunk 2]
```diff
@@ -465,15 +472,50 @@ setupMarloweParams owners contract = mapError (review _MarloweError) $ do
...
+              mustPayToShelleyAddress addr (Val.singleton rolesSymbol role 1 <> minAdaTxOut)
+        distributeRoleTokens <- foldMapM giveToParty $ AssocMap.toList owners
         let params = marloweParams rolesSymbol
         pure (params, mintTx <> distributeRoleTokens, lookups)
     else do
         let missingRoles = roles `Set.difference` Set.fromList (AssocMap.keys owners)
         let message = T.pack $ "You didn't specify owners of these roles: " <> show missingRoles
         throwing _ContractError $ OtherError message
 
+ownShelleyAddress
```
Comment (bwbush) [inline marlowe/src/Language/Marlowe/Client.hs:485]: We might consider a PR to move these into `plutus-apps`, but the change would involve adding a constructor to the constraint type there.

File: marlowe/src/Language/Marlowe/Client.hs [hunk 3]
```diff
@@ -465,15 +472,50 @@ setupMarloweParams owners contract = mapError (review _MarloweError) $ do
...
+      in
+        case Shelley.fromShelleyStakeReference stakeRef of
+          StakeAddressByValue (StakeCredentialByScript _) ->
+            throwError $ OtherContractError $ OtherError "Script stake addresses not supported"
+          StakeAddressByPointer _ ->
+            throwError $ OtherContractError $ OtherError "Pointer stake addresses not supported"
+          NoStakeAddress -> pure (ppkh, Nothing)
+          StakeAddressByValue (StakeCredentialByKey stakeHash) ->
+            pure (ppkh,  Just . StakePubKeyHash . PubKeyHash . toBuiltin $ serialiseToRawBytes stakeHash)
+shelleyAddressToKeys _ = throwError $ OtherContractError $ OtherError "Byron Addresses not supported"
```
Comment (bwbush) [inline marlowe/src/Language/Marlowe/Client.hs:518]: It actually wouldn't be difficult to support Byron addresses, the `Cardano.Api.Byron` module has address functions similar to `Cardano.Api.Shelley`.

**Review by bwbush — APPROVED** [review]
Looks good!

I'll need to adjust some of the still non-automated tests for the PAB, but that will be trivial, and I can do that after this is merged.

File: marlowe/src/Language/Marlowe/Client.hs (hunk 1 above)
Comment (jhbertra) [reply marlowe/src/Language/Marlowe/Client.hs:136]: `AddressInEra` has a `FromJSON` instance, `AddressAny` doesn't

File: marlowe/src/Language/Marlowe/Client.hs (hunk 3 above)
Comment (jhbertra) [reply marlowe/src/Language/Marlowe/Client.hs:518]: I'll merge this PR and we can add it if/when we need it

**Pull request merged**